src/Compilers/CSharp/Test/Semantic/Semantics/RefLocalsAndReturnsTests.cs

```diff
@@ -3152,10 +3152,149 @@ IEnumerable<int> localFunction()
+                    yield return y;
+                    y = ref x;
+                    yield return y;
+                }
+                """;
+
+            CreateCompilation(code, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (14,13): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     ref int y = ref x;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "y").WithArguments("ref and unsafe in async and iterator methods").WithLocation(14, 13));
+
+            var expectedOutput = "56";
+
+            CompileAndVerify(code, expectedOutput: expectedOutput, parseOptions: TestOptions.RegularNext).VerifyDiagnostics();
+            CompileAndVerify(code, expectedOutput: expectedOutput).VerifyDiagnostics();
+        }
+
+        [Fact]
+        public void RefLocal_Iterator_LocalFunction_02()
+        {
+            var code = """
+                using System.Collections.Generic;
+
+                int x = 5;
+
+                IEnumerable<int> func()
+                {
+                    ref int y = ref x;
+                    yield return y;
+                    yield return y;
+                }
+
+                func();
+                """;
+
+            CreateCompilation(code, parseOptions: TestOptions.Regular12).VerifyDiagnostics(
+                // (7,13): error CS8652: The feature 'ref and unsafe in async and iterator methods' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                //     ref int y = ref x;
+                Diagnostic(ErrorCode.ERR_FeatureInPreview, "y").WithArguments("ref and unsafe in async and iterator methods").WithLocation(7, 13));
+
+            var expectedDiagnostics = new[]
+            {
+                // (9,18): error CS9217: A 'ref' local cannot be preserved across 'await' or 'yield' boundary.
+                //     yield return y;
+                Diagnostic(ErrorCode.ERR_RefLocalAcrossAwait, "y").WithLocation(9, 18)
+            };
+
+            CreateCompilation(code, parseOptions: TestOptions.RegularNext).VerifyEmitDiagnostics(expectedDiagnostics);
+            CreateCompilation(code).VerifyEmitDiagnostics(expectedDiagnostics);
+        }
+
+        [Fact]
+        public void RefStruct_Iterator_LocalFunction_01()
+        {
+            var code = """
+                using System;
+                using System.Collections.Generic;
+
+                int x = 5;
+
+                foreach (var z in func())
+                {
+                    Console.Write(z);
+                    x++;
+                }
+
+                IEnumerable<int> func()
+                {
+                    Span<int> y = new(ref x);
+                    yield return y[0];
+                    y = new(ref x);
+                    yield return y[0];
```

Comment:
>yield return y[0]; [](http://example.com/codeflow?start=20&length=18)

This looks interesting. Do we understand why this works? The span is handled specially. What component does that and where?